modules/PowermeterBSM/main/powermeterImpl.cpp

```diff
@@ -0,0 +1,152 @@
+
+//////////////////////////////////////////////////////////////////////
+//
+// module related stuff
+
+void powermeterImpl::init() {
+    EVLOG_verbose << "init: ";
+    m_module_start_timestamp =
+        std::chrono::time_point_cast<std::chrono::seconds>(std::chrono::system_clock::now()).time_since_epoch();
+}
+
+void powermeterImpl::ready() {
+
+    m_watchdog_thread = std::thread(&powermeterImpl::watchdog, this);
+}
+
+std::string powermeterImpl::handle_get_signed_meter_value(std::string& auth_token) {
+
+    // your code for cmd get_signed_meter_value goes here
+
+    try {
+
+        std::scoped_lock lock(mod->get_device_mutex());
+
+        transport::AbstractDataTransport::Spt transport = mod->get_data_transport();
+
+        if ( not transport->trigger_snapshot_generation_BSM_OCMF() )
+            EVLOG_debug << __PRETTY_FUNCTION__ << " trigger for OCMF signed snapshot failed! " << std::endl;
+
+        transport::DataVector data = transport->fetch(known_model::BSM_OCMF_CurrentSnapshot);
+
+        bsm::SignedOCMFSnapshot signed_snapshot(data);
+
+        return signed_snapshot.O();
+
+    } catch (const std::runtime_error& e) {
+        EVLOG_error << __PRETTY_FUNCTION__ << " Error: " << e.what() << std::endl;
+        return "get_signed_meter_value_error";
+    }
+};
+
+//////////////////////////////////////////////////////////////////////
+//
+// module module implementation
+
+void powermeterImpl::watchdog() {
+
+    // poor mans watchdog
+    m_publish_variables_worker_thread = std::thread(&powermeterImpl::worker, this);
+
+    while (true) {
+
+        std::this_thread::sleep_for(mod->get_watchdog_interval());
+
+        if (m_publish_variables_worker_thread.joinable()) {
+            EVLOG_warning << __PRETTY_FUNCTION__ << " worker thread has died. restarting... " << std::endl;
+            m_publish_variables_worker_thread.join();
+            m_publish_variables_worker_thread = std::thread(&powermeterImpl::worker, this);
+        }
+    }
+
+    EVLOG_error << __PRETTY_FUNCTION__ << " watchdog thread terminating, module is dead." << std::endl;
+}
+
+void powermeterImpl::worker() {
+
+    try {
+
+        EVLOG_verbose << __PRETTY_FUNCTION__ << " start " << std::endl;
+
+        mod->dump_config_to_log();
+
+        while (true) {
+
+            std::this_thread::sleep_for(mod->get_update_interval());
+
+            try {
+
+                // lock device mutex
+                std::scoped_lock lock(mod->get_device_mutex());
+
+                transport::AbstractDataTransport::Spt transport = mod->get_data_transport();
+
+                EVLOG_debug << __PRETTY_FUNCTION__ << " wakeup. " << std::endl;
+
+                transport::DataVector data = transport->fetch(known_model::Sunspec_ACMeter);
+
+                sunspec_model::ACMeter acmeter(data);
+                types::powermeter::Powermeter result;
+
+                result.timestamp = 0; // FIXME: current implementation is a float that cant store a 32/31 bit
```

Comment:
Is that solved?

Reply:
No, this is a result of the code generated by ev-cli that renders an interface definition "number" as float. This is kept for backward compatibility.